### PR TITLE
Add Silent Payments Support

### PR DESCRIPTION
**Pull request by macgyver13**
Finally consider this PR ready for review and feedback to determine next steps.

## TODOs:
- ~~verify silent payments labeled address handling~~
- ~~verify multi sig silent payment workflows~~
- ~~address edge cases / attack scenarios~~
- (on hold) test on real hardware - evaluate memory and performance
- (partial) proper UI/UX for Silent Payments - **seeking feedback**

### Updated UX with - silent payment address - appended to - to address -
SHIFT SYM CAPS
SHIFT SYM CAPS
SHIFT SYM CAPS

### Contribute Shares Prompt 
0F056943EDGE

### CODE REVIEW

File: shared/bech32.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
+# from <https://github.com/sipa/bech32/blob/master/ref/python/segwit_addr.py>
```
[the remark this answers is not in the document]
Comment (macgyver13) [reply shared/bech32.py]: Good catch! Thank you for the feedback.

Unfortunately, I could not use any existing encoding functions in ngu so I followed the scheme of adding nip19 encoding to add bip352_encode.  If you have a better idea please let me know.

[the remark this answers is not in the document]
Comment (scgbckbone) [reply shared/bech32.py]: yeah, I completely forgot we only expose segwit_{encode/decode} thru `ngu.codecs` - sorry for unreliable info.

Your approach, adding it to the libngu, is correct.

File: shared/silentpayments.py [hunk 2]
```diff
@@ -125,6 +125,45 @@ def compute_shared_secret_tweak(ecdh_share_bytes, k):
...
 
 
+def compute_silent_payment_spending_privkey(spend_privkey_bytes, sp_tweak_bytes):
+    """
+    Compute private key for spending a silent payment output
+    
+    BIP-352 formula for spending: d_k = d_spend + t_k (mod n)
+    where:
+    - d_spend is the base spend private key
+    - t_k is the shared secret tweak from PSBT_IN_SP_TWEAK
```
Comment (nymius) [inline shared/silentpayments.py]: The above line is stating correctly `d_spend` is "the base spend private key", but "`t_k` is the shared secret tweak" is ruling out the `label` tweak (both hashes have different tags). I guess you were trying to refer to it too, but I would add it explicitly.

**Comment by macgyver13** [comment]
Summary of changes since March update:
- Added bip352_test_vectors.json and test_bip352_vectors.py
- Added BIP376 spend path validation
- Clear sensitive material in Silent Payments workflows
- Added doc/silentpayments.md
- Added bip352 generic format export
- Total integration tests in test_silentpayments.py:
  - test_sp_signing_story
  - test_sp_p2wpkh_input
  - test_sp_p2tr_input
  - test_sp_mixed_inputs
  - test_sp_mixed_sp_spend_and_regular_input
  - test_sp_three_outputs_same_scan_key
  - test_sp_two_outputs_different_scan_keys
  - test_sp_mixed_output_types
  - test_sp_all_owned_multi_input
  - test_sp_partial_owned_coverage_complete
  - test_sp_partial_owned_coverage_incomplete
  - test_sp_partial_owned_coverage_incomplete_refused
  - test_sp_spend_silent_payment_output
  - test_sp_spend_silent_payment_output_with_taproot
  - test_sp_spend_wrong_tweak_rejected
  - test_sp_spend_wrong_purpose_rejected
  - test_sp_spend_wrong_coin_type_rejected
  - test_exit_gracefully_on_sp_validation_failure
  - test_sp_non_sighash_all_rejected
  - test_sp_psbt_v0_rejected
  - test_sp_spend_to_labeled_change_address

## Testing

### Targeted tests
```
cd testing
pytest test_bip352_vectors.py
pytest test_bip375_vectors.py
pytest test_silentpayments.py
pytest test_export.py::test_export_coldcard
```
### General tests
DEFAULT_PYTEST_MARKS = "not onetime and not veryslow and not manual **and not bitcoind**"
python run_sim_tests.py --multiproc --num-proc 6

**Comment by macgyver13** [comment]
Commit `106990f` fixes silent payment output ordering when multiple outputs share the same scan key (labeled  addresses of one recipient).

**Comment by macgyver13** [comment]
Commit e08bd9c modifies the default silent payment signing flow. The wallet coordinator should be responsible for verifying the PSBT and finalizing for broadcast when silent payments are involved.

**Comment by scgbckbone** [comment]
am I the one blocking this with HW tests ? 

Please, open libngu changes as separate libngu PR.

**Comment by macgyver13** [comment]
> am I the one blocking this with HW tests ?

I consider this PR stable and ready for critical review. Confirming this PR on real hardware would be great!

> Please, open libngu changes as separate libngu PR.

I'll get on that. Thank you!

**Comment by macgyver13** [comment]
Last force push makes the following changes:
- update libngu submodule version to match PR
  - reduce bech32 decode character limit to match submodule for memory efficiency
- remove deadcode in dleq.py
- match bip v1.1.1 bip375_test_vectors.json
- add delay to stabilize test_sp_export* tests

**Review by scgbckbone — COMMENTED** [review]
```
diff --git a/shared/manifest.py b/shared/manifest.py
index 6d4b087a..3aaed79a 100644
--- a/shared/manifest.py
+++ b/shared/manifest.py
@@ -17,6 +17,7 @@ freeze_as_mpy('', [
        'desc_utils.py',
        'dev_helper.py',
        'display.py',
+       'dleq.py',
        'drv_entro.py',
        'exceptions.py',
        'export.py',
@@ -52,6 +53,7 @@ freeze_as_mpy('', [
        'selftest.py',
        'serializations.py',
        'sffile.py',
+       'silentpayments.py',
        'ssd1306.py',
        'stash.py',
        'tapsigner.py',
```

Above patch fixes both Mk/Q HW builds. I didn't test the functionality in any way. Upgrade/downgrade works so you shouldn't stay locked if you have some issue in actual feature. 

Any new file that you add to shared needs to go to (specific) manifest(s). Otherwise it will not get bundled into .dfu and you fail with import errors before PIN entry. Simulator works without it and just inject those files.
